mosaic/mosaic.go

```diff
@@ -0,0 +1,423 @@
+type Symbol string
+
+// Symbol types.
+const (
+	AllSymbols     Symbol = "all"
+	HalfSymbols    Symbol = "half"
+	QuarterSymbols Symbol = "quarter"
+)
+
+// Scale represents scale mode that will be used when rendering the image.
+type Scale uint8
+
+// Symbol types.
+const (
+	None Scale = iota
+	Fit
+	Stretch
+	Center
+)
+
+// Render mosaic with default values.
+func Render(img image.Image) string {
+	m := New()
+	return m.Render(img)
+}
+
+// Mosaic is a renderer and its options.
+type Mosaic struct {
+	outputWidth    int    // Output width.
+	outputHeight   int    // Output height (0 for auto).
+	thresholdLevel uint8  // Threshold for considering a pixel as set (0-255).
+	dither         bool   // Enable Dithering (false as default).
+	useFgBgOnly    bool   // Use only foreground/background colors (no block symbols).
+	invertColors   bool   // Invert colors.
+	scale          int    // Scale level
+	symbols        Symbol // Which symbols to use: "half", "quarter", "all".
+}
+
+// New creates and returns a [Renderer].
+func New() Mosaic {
+	return Mosaic{
+		outputWidth:    80,     // Default width.
+		outputHeight:   0,      // Auto height.
+		thresholdLevel: 128,    // Middle threshold.
+		dither:         false,  // Enable dithering.
+		useFgBgOnly:    false,  // Use block symbols.
+		invertColors:   false,  // Don't invert.
+		scale:          1,      // Don't scale.
+		symbols:        "half", // Use half blocks.
+	}
+}
+
+// PixelBlock represents a 2x2 pixel block from the image.
+type pixelBlock struct {
+	Pixels      [2][2]color.Color // 2x2 pixel grid.
+	AvgFg       color.Color       // Average foreground color.
+	AvgBg       color.Color       // Average background color.
+	BestSymbol  rune              // Best matching character.
+	BestFgColor color.Color       // Best foreground color.
+	BestBgColor color.Color       // Best background color.
+}
+
+type shiftable interface {
+	~uint | ~uint16 | ~uint32 | ~uint64
+}
+
+func shift[T shiftable](x T) T {
+	if x > 0xff {
+		x >>= 8
+	}
+	return x
+}
+
+// Scale sets the [ScaleMode] on [Mosaic].
+func (m Mosaic) Scale(scale int) Mosaic {
+	m.scale = scale
+	return m
+}
+
+// IgnoreBlockSymbols set UseFgBgOnly on [Mosaic].
+func (m Mosaic) IgnoreBlockSymbols(fgOnly bool) Mosaic {
+	m.useFgBgOnly = fgOnly
+	return m
+}
+
+// Dither sets the dither level on [Mosaic].
+func (m Mosaic) Dither(dither bool) Mosaic {
+	m.dither = dither
+	return m
+}
+
+// Set ThresholdLevel on Mosaic
+func (m Mosaic) Threshold(threshold uint8) Mosaic {
+	m.thresholdLevel = threshold
+	return m
+}
+
+// Set InvertColors on Mosaic
+func (m Mosaic) InvertColors(invertColors bool) Mosaic {
+	m.invertColors = invertColors
+	return m
+}
+
+// Set OutputWidth on Mosaic
+func (m Mosaic) Width(width int) Mosaic {
+	m.outputWidth = width
+	return m
+}
+
+// Set OutputHeight on Mosaic
+func (m Mosaic) Height(height int) Mosaic {
+	m.outputHeight = height
+	return m
+}
+
+// Set Symbols on Mosaic
```

Comment:
Honestly, we should have a better name for this

```suggestion
// Symbol sets the mosaic symbol type.
```